README.md
# Hazard discrimination summary
R package for estimating the hazard discrimination summary (HDS). HDS is a measure of time-varying prognostic performance. It can be thought of as an incident, time-varying extension of the discrimination slope (Yates 1982), which is perhaps better known as an important part of the integrated discrimination improvement (IDI; Pencina et al. 2008). Alternatively, HDS is a risk-based complement to the incident/dynamic time-dependent AUC (Heagerty and Zheng 2005). Under some circumstances, HDS also has interesting connections to the Cox model partial likelihood. For a detailed overview of HDS, see Liang and Heagerty (2016) and the related discussions and rejoinder.

## Installation
To install through CRAN, use
```r
install.packages("hds")
```

To install the latest (though not necessarily stable) GitHub version, make sure you have `devtools` installed and use
```r
devtools::install_github("liangcj/hds")
```

## Example
A simple example using the Mayo PBC data from the `survival` package demonstrating both `hds` (estimator based on the Cox model) and `hdslc` (more flexible estimator based on the local-in-time Cox model):
```r
head(hds(times = survival::pbc[1:312, 2],
         status = (survival::pbc[1:312, 3]==2)*1,
         m = survival::pbc[1:312, 5]))

hdsres   <- hds(times=pbc5[,1], status=pbc5[,2], m=pbc5[,3:7])
hdslcres <- hdslc(times = pbc5[,1], status=pbc5[,2], m = pbc5[,3:7], h = 730)
Survt    <- summary(survival::survfit(survival::Surv(pbc5[,1], pbc5[,2])~1))
Survtd   <- cbind(Survt$time, c(0,diff(1-Survt$surv)))
tden     <- density(x=Survtd[,1], weights=Survtd[,2], bw=100, kernel="epanechnikov")

par(mar=c(2.25,2.25,0,0)+0.1, mgp=c(1.25,0.5,0))
plot(c(hdslcres[,1], hdslcres[,1]), c(hdslcres[,2] - 1.96*hdslcres[,3],
                                      hdslcres[,2] + 1.96*hdslcres[,3]),
     type="n", xlab="days", ylab="HDS(t)", cex.lab=.75, cex.axis=.75,
     ylim=c(-3,15), xlim=c(0,3650))
polygon(x=c(hdsres[,1], hdsres[312:1,1]), col=rgb(1,0,0,.25), border=NA,
        fillOddEven=TRUE, y=c(hdsres[,2]+1.96*hdsres[,3],
                              (hdsres[,2]-1.96*hdsres[,3])[312:1]))
polygon(x=c(hdslcres[,1], hdslcres[312:1, 1]), col=rgb(0,0,1,.25), border=NA,
        fillOddEven=TRUE, y=c(hdslcres[,2] + 1.96*hdslcres[,3],
                              (hdslcres[,2] - 1.96*hdslcres[,3])[312:1]))
lines(hdsres[,1], hdsres[,2], lwd=2, col="red")
lines(hdslcres[,1], hdslcres[,2], lwd=2, col="blue")
abline(h=1, lty=3)
legend(x=1200, y=14, legend=c("Proportional hazards",
                              "Local-in-time proportional hazards",
                              "Time density"), col=c("red", "blue", "gray"),
       lwd=2, bty="n", cex=0.75)
with(tden, polygon(c(x, x[length(x):1]),
                   c(y*3/max(y)-3.5, rep(-3.5, length(x))),
                   col="gray", border=NA, fillOddEven=TRUE))
```

## References
#### HDS paper, discussions, rejoinder
Liang CJ and Heagerty PJ (2016). A risk-based measure of time-varying prognostic discrimination for survival models. *Biometrics*. [doi:10.1111/biom.12628](https://doi.org/10.1111/biom.12628)

Gerds TA and Schumacher M (2016). Discussion of “A risk-based measure of time-varying prognostic discrimination for survival models,” by C. Jason Liang and Patrick J. Heagerty. *Biometrics*. [doi:10.1111/biom.12629](https://doi.org/10.1111/biom.12629)

Parast L and Rutter CM (2016). Discussion of “A risk-based measure of time-varying prognostic discrimination for survival models,” by C. Jason Liang and Patrick J. Heagerty. *Biometrics*. [doi:10.1111/biom.12630](https://doi.org/10.1111/biom.12630)

Michael H and Tian L (2016). Discussion of “a risk-based measure of time-varying prognostic discrimination for survival models,” by C. Jason Liang and Patrick J. Heagerty. *Biometrics*. [doi:10.1111/biom.12631](https://doi.org/10.1111/biom.12631)

Liang CJ and Heagerty PJ (2016). Rejoinder to discussions on: A risk-based measure of time-varying prognostic discrimination for survival models. *Biometrics*. [doi:10.1111/biom.12632](https://doi.org/10.1111/biom.12632)

#### Time-dependent ROC and AUC papers
Saha-Chaudhuri P and Heagerty PJ (2013). Non-parametric estimation of a time-dependent predictive accuracy curve. *Biostatistics*, 14(1), 42-59. [doi: 10.1093/biostatistics/kxs021](https://doi.org/10.1093/biostatistics/kxs021)

Heagerty PJ and Zheng Y (2005). Survival model predictive accuracy and ROC curves. *Biometrics*, 61: 92–105. [doi:10.1111/j.0006-341X.2005.030814.x](https://doi.org/10.1111/j.0006-341X.2005.030814.x)

Heagerty PJ, Lumley T, and Pepe MS (2000). Time-Dependent ROC curves for censored survival data and a diagnostic marker. *Biometrics*, 56: 337–344. [doi:10.1111/j.0006-341X.2000.00337.x](https://doi.org/10.1111/j.0006-341X.2000.00337.x)

#### Partial extensions of DS and IDI to survival data
Uno H, Tian L, Cai T, Kohane IS, and Wei LJ (2013). A unified inference procedure for a class of measures to assess improvement in risk prediction systems with survival data. *Statistics in Medicine*, 32: 2430–2442. [doi:10.1002/sim.5647](https://doi.org/doi:10.1002/sim.5647)

Chambless LE, Cummiskey CP, and Cui G (2011). Several methods to assess improvement in risk prediction models: Extension to survival analysis. *Statistics in Medicine*, 30: 22–38. [doi:10.1002/sim.4026](https://doi.org/10.1002/sim.4026)

#### Paper that introduced IDI (and also NRI), and original paper that mentioned DS
Pencina MJ, D' Agostino, RB, D' Agostino, RB, and Vasan RS (2008). Evaluating the added predictive ability of a new marker: From area under the ROC curve to reclassification and beyond. *Statistics in Medicine*, 27: 157–172. [doi:10.1002/sim.2929](https://doi.org/10.1002/sim.2929)

Yates JF (1982). External correspondence: Decompositions of the mean probability score. *Organizational Behavior and Human Performance*, 30: 132–156. [doi:10.1016/0030-5073(82)90237-9](https://doi.org/10.1016/0030-5073(82)90237-9)

#### Empirical processes result (used for development of HDS estimator and asymptotics)
van der Vaart AW and Wellner JA (2007). Empirical processes indexed by estimated functions. *IMS Lecture Notes Monograph
Series*, 55: 234–252. [doi:10.1214/074921707000000382](https://doi.org/10.1214/074921707000000382)

#### Thorough study of the Cox model with time-varying coefficients (used for development of HDS estimator and asymptotics)
Tian L, Zucker D, and Wei LJ (2005). On the Cox model with time-varying regression coefficients. *Journal of the American Statistical Association*, 100(469):172-83. [doi:10.1198/016214504000000845](https://doi.org/10.1198/016214504000000845)

Cai Z and Sun Y (2003). Local linear estimation for time-dependent coefficients in Cox’s regression models. *Scandinavian
Journal of Statistics*, 30: 93–111. [doi:10.1111/1467-9469.00320](https://doi.org/10.1111/1467-9469.00320)


NEWS.md
## hds 0.8.0 (2016-12-30; December 30, 2016)

- Made changes to prepare package for initial submission to CRAN. 

- Added references to relevant journal articles in help files and README.md

- Thanks to Bruce Swihart (@swihart) for both motivation and technical pointers on getting everything ready for a CRAN submission!


R/pbc5.R
#' Cleaned up version of the Mayo PBC data.
#'
#' A cleaned up version of the Mayo PBC data from \code{survival::pbc}. Only the
#' first 312 observations are used (the cases who participated in the
#' randomized trial). Only five of the covariates (listed below) are used.
#' Further, two of the covariates were log transformed.
#'
#' @format A data frame with 312 rows and 7 variables:
#' \describe{
#'   \item{time}{follow up time in days}
#'   \item{status}{1=death, 0=censored}
#'   \item{age}{age in years}
#'   \item{edema}{0=no edema, 0.5=untreated or successfully treated, 1=edema
#'     despite diuretic therapy}
#'   \item{bili}{log serum bilirubin level (original value from
#'     \code{survival::pbc} is unlogged)}
#'   \item{albumin}{serum albumin}
#'   \item{protime}{log standardized blood clotting time (original value from
#'     \code{survival::pbc} is unlogged)}
#' }
#' @source Cleaned up version of \code{survival::pbc}
"pbc5"


R/cox_lc.R
## collection of functions that implement estimation of beta(t) and subject
## specific survival (which requires estimation of cumulative baseline hazard)
## of local constant Cox model as specified by

## Cai and Sun (2003, Scan J of Stat) and Tian Zucker Wei (2005, JASA)
## finda(): local constant estimate of beta(t) using Cai and Sun (2003)
## betahatse.fast(): estimate SE of beta(t) using Cai and Sun (2003)
## sssf.fast(): subject specific survival at all m and t values
##              using Tian Zucker Wei (2005)

#' Estimate the time-varying coefficients from a local-in-time Cox model
#'
#' \code{finda} estimates the time-varying coefficients beta(t) at a single time
#' from a local-in-time Cox model. Think of it as a Cox model where the
#' the coefficients are allowed to vary with time. Further details can be found
#' in Cai and Sun (2003) and Tian et al. (2005).
#'
#' The naming of the function \code{finda} stands for "find a(t)", where "a(t)"
#' is the notation used in Cai and Sun (2003) to represent the time-varying
#' Cox model coefficients. We also refer to "a(t)" as "beta(t)" through the documentation.
#'
#' The user typically will not interact with this function, as \code{finda} is
#' wrapped by \code{hdslc}.
#'
#' @param tt Time to estimate beta(t) at
#' @param times A vector of observed follow up times.
#' @param status A vector of status indicators, usually 0=alive, 1=dead.
#' @param covars A matrix or data frame of numeric  covariate values, with a
#'   column for each covariate and each observation is on a separate row.
#' @param start A vector of length p of starting values to be passed to
#'   \code{optim} for the numerical optimization procedure. p is the number of
#'   covariates. Defaults to all zeroes.
#' @param h A single value on the time scale representing the bandwidth to use.
#' @param ... Additional parameters to pass to \code{optim}.
#' @importFrom stats optim
#' @return A vector of length p, where p is the number of covariates. The vector
#' is the estimated beta(t) from the local-in-time Cox model at time \code{tt}.
#' @references Cai Z and Sun Y (2003). Local linear estimation for time-dependent coefficients
#'   in Cox's regression models. \emph{Scandinavian Journal of Statistics}, 30: 93-111.
#'   \href{https://doi.org/10.1111/1467-9469.00320}{doi:10.1111/1467-9469.00320}
#' @references Tian L, Zucker D, and Wei LJ (2005). On the Cox model with time-varying
#'   regression coefficients. \emph{Journal of the American Statistical Association},
#'   100(469):172-83. \href{https://doi.org/10.1198/016214504000000845}{doi:10.1198/016214504000000845}
finda <- function(tt, times, status, covars, start=rep(0, ncol(covars)), h=400,...){
  # MAKE SURE DATA IS SORTED BY TIME
  # find a(t) for local constant Cox regression as per Cai and Sun (2003)
  # returns a(t) at t=tt
  # uses Epanechnikov kernel
  # added gradient function for roughly 10% speedup in numerical fitting
  nn   <- 1-((times-tt)/h)^2 >= 0 # negative values from epanechnikov
  K    <- (1-((times-tt)/h)^2)*0.75/h # epanechnikov kernel
  #  nn   <- (times-tt)>=(-h) & (times-tt)<=h # negative values from rectangular kernel
  #  K    <- 1/(2*h)                          # rectangular kernel
  N    <- length(times)
  llpl <- function(a){
    if(is.null(ncol(covars))) lp <- covars*a
    else           lp <- covars %*% a
    -sum((K*status*(lp - log(cumsum(exp(lp)[N:1])[N:1])))[nn])
  }
  llplg <- function(a){
    # gradient function
    if(is.null(ncol(covars))) lp <- covars*a
    else           lp <- covars %*% a
    elp <- exp(lp)
    num <- apply(sweep(covars, 1, elp, "*"), 2, function(x) rev(cumsum(rev(x))))
    den <- cumsum(elp[N:1])[N:1] %o% rep(1, ncol(covars))
    #-colSums((K*status*(covars - num/den))[nn,])
    #-colSums(sweep((covars - num/den), 1, K*status, "*")[nn,])
    -colSums(((covars - num/den) * (K*status))[nn, , drop=FALSE])
  }
  #  fit <- optim(par=start, fn=llpl, control=list(maxit=2000), ...)
  fit <- optim(par=start, fn=llpl, gr = llplg, method="BFGS", ...)
  #  cat(fit$convergence, sum(nn), "\n")
  return(fit$par)
}

#' Calculate estimates of the covariance matrix for beta(t).
#'
#' To help with overall computational efficiency of \code{hdslc}, this function
#' will return multiple covariance matrices - one covariance matrix for each
#' requested evaluation time.
#'
#' See Cai and Sun (2003, Scandinavian Journal of Statistics) for details on the
#' local constant estimator for beta(t). The user will not typically interact
#' with this function. The function is wrapped by \code{hdslc}, and the output
#' is used by \code{hdslcse.fast}.
#'
#' @return An N x p x p matrix, where N is the number of evaluation times and
#'   p is the number of covariates.
#' @keywords internal
#' @importFrom tensor tensor
betahatse.fast <- function(betahat, times, status, m, h, evalt){
  n  <- length(status)
  evaln <- length(evalt)
  #evalt <- round(seq(1, n, length.out=100))
  #betahat <- betahat[rep(1:100, c(diff(evalt), 1)), ]
  betahat <- betahat[rep(1:evaln, diff(c(0,evalt[1:(evaln-1)],n))), ]
  ebmm <- exp(tcrossprod(betahat, m))
  S0 <- apply(matrix(1:n), 1, function(x) sum(ebmm[x,x:n]))/n
  S1 <- matrix(NA, nrow=n, ncol=ncol(m))
  for(i in 1:n){S1[i,] <- (ebmm[i,i:n, drop=FALSE] %*% m[i:n, , drop=FALSE])/n}
  # can speed up by writing C code that does this
  mouter <- array(NA, dim=c(dim(m), ncol(m)))
  for(i in 1:n){mouter[i, , ] <- m[i, ] %o% m[i, ]}
  S2 <- array(NA, dim=c(n, ncol(m), ncol(m)))
  for(i in 1:n){S2[i, , ] <- tensor::tensor(ebmm[i,i:n, drop=FALSE], mouter[i:n, , , drop=FALSE], 2, 1)/n}
  A <- sweep(S2, 1, S0, "/")
  Btmp <- sweep(S1, 1, S0, "/")
  B <- array(NA, dim=c(dim(m), ncol(m)))
  for(i in 1:n){B[i, , ] <- outer(Btmp[i, ], Btmp[i, ])}
  V <- A-B
  tmat <- apply(matrix(times[evalt]), 1, function(x) times-x)
  # each i-th column is the n times minus the i-th time
  K <- ((1-((tmat)/h)^2)*0.75/h)*(1-((tmat)/h)^2 >= 0) # epanechnikov kernel
  integ <- tensor::tensor(V, K, 1, 1)/n
  I <- array(NA, dim=c(evaln, ncol(m), ncol(m)))
  for(i in 1:evaln){
    try(I[i, , ] <- solve(integ[ , , i]))
  }
  return(I * 0.6)
}

sssf.fast <- function(betat, status, m, evalt){
  # calculates subject specific survival for all t and all m values
  # faster, vectorized version
  n <- length(status)
  evaln <- length(evalt)
  #evalt <- round(seq(1, n, length.out=100))
  etbm <- exp(tcrossprod(betat, m))
  exp(-apply(matrix(1:n), 1, function(x){
    etbmn <- etbm / etbm[, x]
    #cumsum(status * (1/apply(matrix(1:n), 1, function(y) sum(etbmn[y, y:n]))))
    doit <- rep(NA, evaln)
    #for(i in 1:99){
    #  tmp <- sum(etbmn[i, (evalt[i]):n]) - c(0, cumsum(etbmn[i, (evalt[i]+1):(evalt[i+1]-1)]))
    #  doit[i] <- (1/tmp) %*% status[evalt[i]:(evalt[i+1]-1)]
    #}
    #doit[100] <- status[n] * (1/etbmn[i, n])
    for(i in 2:evaln){
      if(evalt[i]==0) doit[i] <- NA
      else{
        if(evalt[i]==1) doit[i] <- status[1] * (1/sum(etbmn[1, ]))
        else{
          if((evalt[i] == evalt[i-1]) | (evalt[i] == evalt[i-1]+1)) doit[i] <- (1/sum(etbmn[i, (evalt[i]):n])) * status[evalt[i]]
          else{
            tmp <- sum(etbmn[i, (evalt[i]):n]) + c(rev(cumsum(etbmn[i, (evalt[i]-1):(evalt[i-1]+1)])), 0)
            doit[i] <- (1/tmp) %*% status[(evalt[i-1]+1):(evalt[i])]
          }
        }
      }
    }
    if(evalt[1]==0) doit[1] <- NA
    else{
      if(evalt[1]==1) doit[1] <- (1/sum(etbmn[1, ])) * status[1]
      else{
        tmp <- sum(etbmn[1, (evalt[1]):n]) + c(rev(cumsum(etbmn[1, (evalt[1]-1):1])), 0)
        doit[1] <- (1/tmp) %*% status[1:(evalt[1])]
      }
    }
    #doit[1] <- (1/sum(etbmn[1, ])) * status[1]
    cumsum(doit)
  }))
}



R/hds_lc_hat.R
#### functions for calculating local constant HDS(t) and its standard errors
#### functions for calculating beta(t), Lambda(t), and beta(t) SE in cox_lc.R
#### local constant HDS(t) SE basically just depends on beta(t) SE

#' Hazard discrimination summary estimate (local constant) at one time point
#'
#' \code{hdslc.fast} estimates HDS at a single time using the local-in-time
#' proportional hazards model. See Cai and Sun (2003, Scandinavian Journal of
#' Statistics) for details on the local-in-time PH model.
#'
#' The user typically will not interact with this function. Rather, \code{hdslc}
#' wraps this function and is what the user typically will use.
#'
#' @param S A vector of length \code{nrow(m)} (which is typically the number of
#' observations n), where each value is the subject-specific survival at time t
#' where t is implied by the choice of \code{betahat}.
#' @param betahat A p x 1 vector of coefficient estimates at time t of interest
#'   from the local-in-time Cox model. Vector length p is the number of
#'   covariates. Typically the output from \code{hdslc::finda} is passed here.
#' @param m A numeric n x p matrix of covariate values, with a column for each
#'   covariate and each observation is on a separate row.
#' @return The HDS estimate at times t, where t is implied by choice of \code{S}
#' and \code{betahat} passed to \code{hdslc.fast}.
hdslc.fast <- function(S, betahat, m){
  if(is.null(dim(m))) m <- matrix(m)
  else                m <- matrix(m, ncol=ncol(m))
  bm     <- m%*%betahat
  S3hat  <- sum(exp(2*bm)*S)
  S1hat  <- sum(S)
  S2hat  <- sum(exp(bm)*S)
  return(S3hat*S1hat/(S2hat^2))
}

#' Hazard discrimination summary (local constant) standard error estimate
#'
#' \code{hdslcse.fast} calculates an estimate of the variance for the
#' local constant hazard discrimination summary estimator at a time t. The time
#' t is implied by \code{S}, \code{betahat}, and \code{betahatse}
#'
#' The use will typically not interact with this function directly. Instead this
#' function is wrapped by \code{hdslc}.
#'
#' @param S A vector of length \code{nrow(m)} (which is typically the number of
#' observations n), where each value is the subject-specific survival at time t
#' where t is implied by the choice of \code{betahat}.
#' @param betahat A p x 1 vector of coefficient estimates at time t of interest
#'   from the local-in-time Cox model. Vector length p is the number of
#'   covariates. Typically the output from \code{hdslc::finda} is passed here.
#' @param m A numeric n x p matrix of covariate values, with a column for each
#'   covariate and each observation is on a separate row.
#' @param betahatse A p x p covariance matrix for betahat at time t
#' @return Variance estimate that has not been normalized. To get a usable
#'   standard error estimate, divide the output of this function by the
#'   bandwidth and sample size, and then take the square root.
hdslcse.fast <- function(S, betahat, m, betahatse){
  if(is.null(dim(m))) m <- matrix(m)
  else                m <- matrix(m, ncol=ncol(m))
  n    <- nrow(m)
  bm   <- m %*% betahat
  bmm  <- bm %*% rep(1, ncol(m))
  Smm  <- S %o% rep(1, ncol(m))

  S2hat <- colSums(2*m*exp(2*bmm)*Smm)/n
  S1hat <- colSums(m*exp(bmm)*Smm)/n
  #S0hat <- colSums(Smm)/n
  S0hat <- rep(0, ncol(m))

  Smat  <- rbind(S0hat, S1hat, S2hat)
  Sigma2 <- Smat %*% betahatse %*% t(Smat)

  Pf2  <- sum(exp(2*bm)*S)/n
  Pf0  <- sum(S)/n
  Pf1  <- sum(exp(bm)*S)/n

  A <- matrix(c(Pf2/(Pf1^2), -(2*Pf2*Pf0)/(Pf1^3), Pf0/(Pf1^2)), nrow=1)
  return(A %*% Sigma2 %*% t(A))
}

#' Hazard discrimination summary estimator
#'
#' Returns local constant HDS estimates at all specified evaluation times.
#' See Liang and Heagerty (2016) for details on HDS.
#'
#' A local constant version of \code{hds}. While \code{hds} estimates HDS(t)
#' assuming the Cox proportional hazards model, \code{hdslc} estimates HDS(t)
#' using a relaxed, local-in-time Cox model. Specifically, the hazard
#' ratios are allowed to vary with time. See Cai and Sun (2003) and Tian Zucker Wei (2005) for details on the
#' local-in-time Cox model.
#'
#' Point estimates use \code{hdslc.fast} and standard errors use
#' \code{hdslcse.fast}. \code{hdslc.fast} requires an estimate of beta(t) (time-varying
#' hazard ratio), which is estimated using \code{finda()}; and subject specific
#' survival, which is estimated using sssf.fast(). \code{hdslcse.fast} requires the same
#' and in addition standard error estimates of beta(t), which are estimated
#' using \code{betahatse.fast()}.
#'
#' The covariate values \code{m} are centered for numerical stability. This is
#' particularly relevant for the standard error calculations.
#'
#' @param times A vector of observed follow up times.
#' @param status A vector of status indicators, usually 0=alive, 1=dead.
#' @param m A matrix or data frame of covariate values, with a column for each
#'   covariate and each observation is on a separate row. Non-numeric values
#'   are acceptable, as the values will be transformed into a numeric model
#'   matrix through \code{survival::coxph}.
#' @param evaltimes A vector of times at which to estimate HDS. Defaults to all
#'   the times specified by the \code{times} vector. If there are a lot of
#'   observations, then you may want to enter in a sparser vector of times for
#'   faster computation.
#' @param h A single numeric value representing the bandwdith to use, on the
#'   time scale. The default bandwidth is a very ad hoc estimate using
#'   Silverman's rule of thumb
#' @param se TRUE or FALSE. TRUE: calculate and return standard error estimates.
#'   FALSE: do not calculate standard errors estimates and return NAs. Defaults
#'   to TRUE. May want to set to FALSE to save computation time if using this
#'   function to compute bootstrap standard errors.
#' @references Liang CJ and Heagerty PJ (2016).
#'   A risk-based measure of time-varying prognostic discrimination for survival
#'   models. \emph{Biometrics}. \href{https://doi.org/10.1111/biom.12628}{doi:10.1111/biom.12628}
#' @references Cai Z and Sun Y (2003). Local linear estimation for time-dependent coefficients
#'   in Cox's regression models. \emph{Scandinavian Journal of Statistics}, 30: 93-111.
#'   \href{https://doi.org/10.1111/1467-9469.00320}{doi:10.1111/1467-9469.00320}
#' @references Tian L, Zucker D, and Wei LJ (2005). On the Cox model with time-varying
#'   regression coefficients. \emph{Journal of the American Statistical Association},
#'   100(469):172-83. \href{https://doi.org/10.1198/016214504000000845}{doi:10.1198/016214504000000845}
#' @seealso \code{\link{hds}}, \code{\link{finda}}
#' @examples
#' \dontrun{
#' head(hdslc(times = survival::pbc[1:312, 2],
#'            status = (survival::pbc[1:312, 3]==2)*1,
#'            m = survival::pbc[1:312, 5]))
#'
#' hdsres   <- hds(times=pbc5[,1], status=pbc5[,2], m=pbc5[,3:7])
#' hdslcres <- hdslc(times = pbc5[,1], status=pbc5[,2], m = pbc5[,3:7], h = 730)
#' Survt    <- summary(survival::survfit(survival::Surv(pbc5[,1], pbc5[,2])~1))
#' Survtd   <- cbind(Survt$time, c(0,diff(1-Survt$surv)))
#' tden     <- density(x=Survtd[,1], weights=Survtd[,2], bw=100, kernel="epanechnikov")
#'
#' par(mar=c(2.25,2.25,0,0)+0.1, mgp=c(1.25,0.5,0))
#' plot(c(hdslcres[,1], hdslcres[,1]), c(hdslcres[,2] - 1.96*hdslcres[,3],
#'                                       hdslcres[,2] + 1.96*hdslcres[,3]),
#'      type="n", xlab="days", ylab="HDS(t)", cex.lab=.75, cex.axis=.75,
#'      ylim=c(-3,15), xlim=c(0,3650))
#' polygon(x=c(hdsres[,1], hdsres[312:1,1]), col=rgb(1,0,0,.25), border=NA,
#'         fillOddEven=TRUE, y=c(hdsres[,2]+1.96*hdsres[,3],
#'                               (hdsres[,2]-1.96*hdsres[,3])[312:1]))
#' polygon(x=c(hdslcres[,1], hdslcres[312:1, 1]), col=rgb(0,0,1,.25), border=NA,
#'         fillOddEven=TRUE, y=c(hdslcres[,2] + 1.96*hdslcres[,3],
#'                               (hdslcres[,2] - 1.96*hdslcres[,3])[312:1]))
#' lines(hdsres[,1], hdsres[,2], lwd=2, col="red")
#' lines(hdslcres[,1], hdslcres[,2], lwd=2, col="blue")
#' abline(h=1, lty=3)
#' legend(x=1200, y=14, legend=c("Proportional hazards",
#'                               "Local-in-time proportional hazards",
#'                               "Time density"), col=c("red", "blue", "gray"),
#'        lwd=2, bty="n", cex=0.75)
#' with(tden, polygon(c(x, x[length(x):1]),
#'                    c(y*3/max(y)-3.5, rep(-3.5, length(x))),
#'                    col="gray", border=NA, fillOddEven=TRUE))
#' }
#'
#' @return A data frame with three columns: 1) the evaluation times, 2) the HDS
#'   estimates at each evaluation time, and 3) the standard error estimates at
#'   each evaluation time
#' @export
#' @importFrom survival Surv coxph
#' @importFrom stats sd
hdslc <- function(times,
                  status,
                  m,
                  evaltimes=times[order(times)],
                  h=1.06*sd(times)*(length(times)^(-0.2)),
                  se=TRUE){
  m   <- as.matrix(m)
  fit <- coxph(Surv(times, status)~m, x = TRUE)
  m   <- fit$x
  m   <- scale(m, scale = FALSE)     # center marker values
  m   <- as.matrix(m)

  timesort   <- order(times)     # this and next three lines sort data by time
  m_s        <- m[timesort, , drop=FALSE]    # sorted marker values
  status_s   <- status[timesort] # sorted status values
  times_s    <- times[timesort]  # sorted times

  evalt <- findInterval(evaltimes, times_s) # find indices of evaluation times
  evaln <- length(evalt)                    # how many evaluation times

  betaD <- apply(matrix(times_s[evalt]), 1, function(x)
    finda(x, times=times_s, status=status_s, covars=m_s, h=h))
  betaD <- t(betaD)
  if(nrow(betaD)==1) betaD <- t(betaD)
  es <- sssf.fast(betat = betaD, status = status_s, m = m_s, evalt)
  hdslcres <- apply(matrix(1:evaln), 1, function(x)
    hdslc.fast(es[x, ], betaD[x, ], m_s))
  hdslcseres <- rep(NA, evaln)
  if(se){
    # calculate standard errors if se=TRUE
    betase <- betahatse.fast(betahat = betaD, times = times_s, status = status_s,
                             m = m_s, h = h, evalt)
    hdslcseres <- apply(matrix(1:evaln), 1, function(x){
      hdslcse.fast(es[x,], betaD[x, ], m_s, betase[x,,])})
    hdslcseres <- sqrt(hdslcseres/(h*length(times_s)))
  }
  return(data.frame(times=evaltimes, hdslchat=hdslcres, se=hdslcseres))
}

